cmd/utils/cmd.go

```diff
@@ -185,32 +190,87 @@ func ExportChain(blockchain *core.BlockChain, fn string) error {
 		defer writer.(*gzip.Writer).Close()
 	}
 
-	if err := blockchain.Export(writer); err != nil {
+	var z *zip.Writer
+	if strings.HasSuffix(fn, ".zip") {
+		z = zip.NewWriter(fh)
+		writer, err = z.Create("blocks.dat")
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Info("Writing blocks", "first", first, "last", last)
+	if err := blockchain.ExportN(writer, first, last); err != nil {
 		return err
 	}
-	log.Info("Exported blockchain", "file", fn)
 
+	if z != nil {
+		log.Info("Writing state", "blockNumber", last)
+		writer, err := z.Create("state.dat")
+		if err != nil {
+			return err
+		}
+		if err := exportState(blockchain, chainDb, writer, last); err != nil {
+			return err
+		}
+
+		log.Info("Writing receipts", "blockNumber", last)
+		writer, err = z.Create("receipts.dat")
+		if err != nil {
+			return err
+		}
+		if err := exportReceipts(blockchain, chainDb, writer, last); err != nil {
+			return err
+		}
+
+		err = z.Close()
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Info("Exported blockchain to", "file", fn)
 	return nil
 }
 
-func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, last uint64) error {
-	log.Info("Exporting blockchain", "file", fn)
-	// TODO verify mode perms
-	fh, err := os.OpenFile(fn, os.O_CREATE|os.O_APPEND|os.O_WRONLY, os.ModePerm)
+func exportState(blockchain *core.BlockChain, chainDb ethdb.Database, writer io.Writer, blocknum uint64) error {
+	sdb, err := blockchain.StateAt(blockchain.GetHeaderByNumber(blocknum).Root)
 	if err != nil {
 		return err
 	}
-	defer fh.Close()
 
-	var writer io.Writer = fh
-	if strings.HasSuffix(fn, ".gz") {
-		writer = gzip.NewWriter(writer)
-		defer writer.(*gzip.Writer).Close()
+	iter := state.NewNodeIterator(sdb)
+	nodes, hashnodes := 0, 0
+	for iter.Next() {
+		nodes += 1
+		if nodes%100000 == 0 {
+			log.Info("Exported nodes", "count", nodes, "hashnodes", hashnodes)
+		}
+		if iter.Hash != (common.Hash{}) {
+			hashnodes += 1
+			entry, err := chainDb.Get(iter.Hash.Bytes())
+			if err != nil {
+				return err
+			}
+			writer.Write(entry)
+		}
 	}
 
-	if err := blockchain.ExportN(writer, first, last); err != nil {
-		return err
+	return nil
+}
+
+func exportReceipts(blockchain *core.BlockChain, chainDb ethdb.Database, writer io.Writer, blocknum uint64) error {
+	count := 0
+	var i uint64
+	for i = 0; i <= blocknum; i++ {
+		receipts := core.GetBlockReceipts(chainDb, blockchain.GetHeaderByNumber(i).Hash(), i)
+		count += len(receipts)
+		for _, receipt := range receipts {
+			receipt.EncodeRLP(writer)
```

Comment:
Are you sure this is correct? How will you know during importing which receipts go into this block and which go into the next? Why not encode all receipts as one RLP?

Also please add error handling. You may run out of disk, in which case this method will fail.